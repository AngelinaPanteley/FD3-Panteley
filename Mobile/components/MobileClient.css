.MobileClient
{
  padding: 2px 0;
  display: flex;
  align-items: center;
}

.MobileClientBalance
{
  width: 100px;
  display: inline-flex;
  align-items: center;
  justify-content: space-between;
}

.MobileClientFio
{
  display: inline-flex;
  width: 300px;
  padding-right: 20px;
  justify-content: space-between;
  box-sizing: border-box;
  align-items: center;
}

.MobileClientFioForm {
  display: inline-flex;
}

.MobileClientFioForm input {
  width: 30%;
}

.MobileClientFioContainer {
  display: inline-flex;
  justify-content: space-between;
  width: 100%;
}

.MobileClientActions {
  display: inline-block;
  margin-left: 20px;
}

.MobileClientActions button {
  margin: 0 5px;
}

.MobileClientStatus {
  display: inline-block;
  margin-left: 25px;
  width: 50px;
}

.MobileClientStatusActive {
  color: green;
}

.MobileClientStatusBlocked {
  color: red;
}

.MobileClientId {
  display: inline-block;
  margin-top: -6px;
  margin-right: 10px;
}