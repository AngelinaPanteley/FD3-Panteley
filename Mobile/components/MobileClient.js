import React from 'react';
import PropTypes from 'prop-types';

import './MobileClient.css';

class MobileClient extends React.PureComponent {

  static propTypes = {
    client: PropTypes.shape({
      id: PropTypes.number.isRequired,
      fam: PropTypes.string.isRequired,
      im: PropTypes.string.isRequired,
      otch: PropTypes.string.isRequired,
      balance: PropTypes.number.isRequired,
      isEdit: PropTypes.bool.isRequired,
    }),
    onDelete: PropTypes.func.isRequired,
    onFIOChange: PropTypes.func.isRequired,
    onBalanceChange: PropTypes.func.isRequired,
  };

  state = {
    fam: this.props.client.fam,
    im: this.props.client.im,
    otch: this.props.client.otch,
    balance: this.props.client.balance,
    isActive: this.props.client.balance >= 0,
    isEdit: this.props.client.isEdit,
  };

  componentWillReceiveProps = (newProps) => {
    const {fam, im, otch} = this.state;
    const newClient = newProps.client;

    if (fam !== newClient.fam || im !== newClient.im || otch !== newClient.otch) {
      this.setState({
        fam: newClient.fam,
        im: newClient.im,
        otch: newClient.otch,
        isEdit: false,
      });
    }

    if (this.state.balance !== newClient.balance) {
      this.setState({
        balance: newClient.balance,
        isActive: newClient.balance >= 0,
      });
    }

    if (this.state.isEdit !== newClient.isEdit) {
      this.setState({
        isEdit: newClient.isEdit,
      });
    }
  };

  deleteHandle = () => {
    this.props.onDelete(this.props.client.id);
  }

  editButtonHandle = () => {
    this.setState({ isEdit: true });
  }

  increaseBalance = () => {
    this.props.onBalanceChange(this.props.client.id, this.state.balance + 1);
  }

  decreaseBalance = () => {
    this.props.onBalanceChange(this.props.client.id, this.state.balance - 1);
  }

  saveForm = (e) => {
    e.preventDefault();
    const form = e.target;
    const FIO = {
      fam: form.fam.value,
      im: form.im.value,
      otch: form.otch.value,
    };
    this.props.onFIOChange(this.props.client.id, FIO);
  }

  render() {
    console.log("MobileClient id=" + this.props.client.id + " render");

    const {fam, im, otch} = this.state;

    return (
      <div>
        <div className='MobileClient'>
          <span className="MobileClientId">{this.props.client.id}</span>
          <div className='MobileClientFio'>
            {
              this.state.isEdit
                ?
                <form onSubmit={this.saveForm} className="MobileClientFioForm">
                  <input type="text" defaultValue={fam} name="fam" />
                  <input type="text" defaultValue={im} name="im" />
                  <input type="text" defaultValue={otch} name="otch" />
                  <input type="submit" value="Save" />
                </form>
                :
                <div className="MobileClientFioContainer">
                  <span>{fam + " " + im + " " + otch}</span>
                  <button onClick={this.editButtonHandle}>Edit</button>
                </div>
            }
          </div>
          <div className='MobileClientBalance'>
            <button onClick={this.decreaseBalance}>-</button>
            <span>{this.state.balance}</span>
            <button onClick={this.increaseBalance}>+</button>
          </div>
          <div className='MobileClientStatus'>
            {
              this.state.isActive
                ?
                <span className='MobileClientStatusActive'>Active</span>
                :
                <span className='MobileClientStatusBlocked'>Blocked</span>
            }
          </div>
          <div className='MobileClientActions'>
            <button onClick={this.deleteHandle}>Delete</button>
          </div>
        </div>
      </div>
    );

  }

}

export default MobileClient;
