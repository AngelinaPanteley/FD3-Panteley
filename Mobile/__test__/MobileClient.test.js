"use strict";

import React from 'react';
import renderer from 'react-test-renderer';
import { expect as chaiExpect} from 'chai';
import { shallow, configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16.1';

import MobileClient from '../components/MobileClient.js';

configure({ adapter: new Adapter() });

describe('>>>MobileClient', () => {

  //SNAPSHOTS

  const client = {
    id: 1,
    fam: 'some fam',
    im: 'some im',
    otch: 'some otch',
    balance: 0,
    isEdit: false,
  };

  const clientComponent = <MobileClient client={client} onDelete={() => { }}
    onFIOChange={() => {}}
    onBalanceChange={() => {}} />;

  const component = renderer.create(clientComponent);

  it('renders client correctly', () => {
    let componentTree = component.toJSON();
    expect(componentTree).toMatchSnapshot();
  });

  const client2 = {
    id: 222,
    fam: 'some fam2',
    im: 'some im2',
    otch: 'some otch2',
    balance: 222,
    isEdit: true,
  };

  const clientComponent2 = <MobileClient client={client2} onDelete={() => { }}
    onFIOChange={() => {}}
    onBalanceChange={() => {}} />;

  const component2 = renderer.create(clientComponent2);

  it('renders editable client correctly', () => {
    let componentTree = component2.toJSON();
    expect(componentTree).toMatchSnapshot();
  });

  //SHALLOW RENDERING

  let wrapper;
  
  beforeEach(() => {
    wrapper = shallow(clientComponent);
  });

  it('renders client`s id', () => {
    chaiExpect(wrapper.contains(<span className="MobileClientId">{client.id}</span>)).to.equal(true);
  });

  it('renders client`s fio', () => {
    chaiExpect(wrapper.find('.MobileClientFio')).to.have.lengthOf(1);
  });

  it('renders client`s balance', () => {
    chaiExpect(wrapper.find('.MobileClientBalance')).to.have.lengthOf(1);
  });

  it('renders client`s status', () => {
    chaiExpect(wrapper.find('.MobileClientStatus')).to.have.lengthOf(1);
  });

});
