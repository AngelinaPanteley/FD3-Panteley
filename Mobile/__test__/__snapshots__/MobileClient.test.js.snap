// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`>>>MobileClient renders client correctly 1`] = `
<div
  className="MobileClient"
>
  <span
    className="MobileClientId"
  >
    1
  </span>
  <div
    className="MobileClientFio"
  >
    <div
      className="MobileClientFioContainer"
    >
      <span>
        some fam some im some otch
      </span>
      <button
        onClick={[Function]}
      >
        Edit
      </button>
    </div>
  </div>
  <div
    className="MobileClientBalance"
  >
    <button
      className="MobileClientDecreaseBalance"
      onClick={[Function]}
    >
      -
    </button>
    <span>
      0
    </span>
    <button
      className="MobileClientIncreaseBalance"
      onClick={[Function]}
    >
      +
    </button>
  </div>
  <div
    className="MobileClientStatus"
  >
    <span
      className="MobileClientStatusActive"
    >
      Active
    </span>
  </div>
  <div
    className="MobileClientActions"
  >
    <button
      onClick={[Function]}
    >
      Delete
    </button>
  </div>
</div>
`;

exports[`>>>MobileClient renders editable client correctly 1`] = `
<div
  className="MobileClient"
>
  <span
    className="MobileClientId"
  >
    222
  </span>
  <div
    className="MobileClientFio"
  >
    <form
      className="MobileClientFioForm"
      onSubmit={[Function]}
    >
      <input
        defaultValue="some fam2"
        name="fam"
        type="text"
      />
      <input
        defaultValue="some im2"
        name="im"
        type="text"
      />
      <input
        defaultValue="some otch2"
        name="otch"
        type="text"
      />
      <input
        type="submit"
        value="Save"
      />
    </form>
  </div>
  <div
    className="MobileClientBalance"
  >
    <button
      className="MobileClientDecreaseBalance"
      onClick={[Function]}
    >
      -
    </button>
    <span>
      222
    </span>
    <button
      className="MobileClientIncreaseBalance"
      onClick={[Function]}
    >
      +
    </button>
  </div>
  <div
    className="MobileClientStatus"
  >
    <span
      className="MobileClientStatusActive"
    >
      Active
    </span>
  </div>
  <div
    className="MobileClientActions"
  >
    <button
      onClick={[Function]}
    >
      Delete
    </button>
  </div>
</div>
`;
